src/mem/Allocator.h

```diff
@@ -54,13 +72,62 @@ class Allocator : public Interface
     /// XXX: Misplaced -- not all allocators have a notion of a "chunk". See MemPoolChunked.
     virtual void setChunkSize(size_t) {}
 
+    virtual bool idleTrigger(int shift) const = 0;
+
+    virtual void clean(time_t maxage) = 0;
+
+    /**
+     * Flush temporary counter values into the statistics held in 'meter'.
+     * Performs some CPU intensive calculations which have been deferred for performance.
```

Comment:
```C++
void update(size_t items, size_t itemSize) {
    count += items;
    bytes += (items * itemSize);
}
```

The two increments an one multiplication above are hardly "CPU intensive calculations" worth "deferring for performance", especially since the bytes update is not really needed until the meters are aggregated and/or reported. Are you sure all this complexity is added for performance reasons?

Reply:
AFAICT these counters elide 4*999 additions of objectSize(), 999 virtual table lookups of getMeter(), 2*999 pointer dereference calculations, and lag time loading the separate 'meter' object into CPU cache 999 times. The "fast" counters are members of the Allocator which puts them already in CPU memory as side effects of loading it for the unavoidable part of the allocation. At cost of 3 extra multiplications every 1000 allocation events.

So yeah, I'm pretty sure it is there for performance. If you think it is *not enough* gain, we can add a few more zeros to FlushLimit. I would rather not touch it until the membanger tests are refreshed.

Reply:
> elide 4x999 additions of objectSize()

Where would we add objectSize() on every alloc()/freeOne() if, for example, these "fast" counters become a part of "slow" Meter or PoolMeter, removing FastCounters? As detailed elsewhere, byte calculations are indeed unnecessary until meters are merged or printed. We do not need to update a `gb_*` field on every alloc()/freeOne() call, regardless of FastCounters existence.

> 1x999 virtual table lookups of getMeter()

The `meter` is a member of this class so no virtual table lookups are necessary to get to "slow" Meter or PoolMeter fields.

> 2x999 pointer dereference calculations, and lag time loading the separate 'meter' object into CPU cache 999 times

The `meter` is a data member of this Allocator class, just like `counter` is. It is not more or less "separate" than `counter`.

In summary, all these claims of CPU optimizations seem to be gross exaggerations -- about the same amount of calculations would be performed if FastCounter did not exist (with its fields merged into Meter and/or PoolMeter, as needed).

Please note that this change request is not asking to remove FastCounter. At this point, I am only trying to establish the facts. We will decide what the best solution is after we understand what is actually going on in this code.

Reply:
> At this point, I am only trying to establish the facts. We will decide what the best solution is after we understand what is actually going on in this code.

Ah, "what" is an entirely different topic than "justify this".

The facts known are the documentation comment I wrote. These counters are a pre-existing performance optimization - simply that. Whether it is actually useful or not requires testing I am not about to do here - my understanding of its reason are as stated above.

If you want a better comment text, sure. But I am not intending to polish the logic of optimizations **right now**, halfway through a series of refactors, especially ones that I am not certain of the specific performance benefit/loss.

Reply:
> The facts known are the documentation comment I wrote.

The proposed documentation relies on an claim that these counters/actions are a performance optimization. There is currently no (shared) knowledge of that claim validity. In other words, that alleged fact has not been established yet. The already written documentation does not support that fact yet (due to problems detailed elsewhere).

> These counters are a pre-existing performance optimization

I see no evidence of those counters being (as in "existing to provide") a performance optimization.

If you prefer, we can ignore the lack of shared understanding regarding performance optimization status of these counters. AFAICT, that shared understanding is _not_ necessary to merge this PR. We will just need to remove some comments and adjust some names. Please let me know if I should switch to that strategy (instead of asking questions about these counters/actions in hope to correct their documentation/names).

Reply:
> The proposed documentation relies on an claim that these counters/actions are a performance optimization. There is currently no (shared) knowledge of that claim validity. In other words, that alleged fact has not been established yet.

AFAICS we have both stated that there is some (non-zero) performance gain from them.

Are you aware of any raison d'être of the variables this PR puts into FastCounters which is not about performance?
If not please accept the claim so we can move forward.

Reply:
> Are you aware of any raison d'être of the variables this PR puts into FastCounters which is not about performance?

I am. Based on that knowledge, I have now suggested specific FastCounters changes in the change request about that class.

As far as _this_ change request is concerned, I suggest simply removing the second sentence of the method description. To move forward, we do not need to agree on the performance impact of this method (as long as neither of the conflicting opinions is documented as the established fact).

Reply:
> > Are you aware of any raison d'être of the variables this PR puts into FastCounters which is not about performance?
> 
> I am. 

Please share technical detail/reasons instead of obtusely requiring design changes.

Reply:
> Please share technical detail/reasons 

I have done that and even provided (a link to) those technical details/reasons expressed as C++ code. In summary, these data members count the number of various events since the last MemImplementingAllocator flushing event. These data members exist to provide that statistics. They are problematic, on several levels, but we do not need to agree on the exact set of those problems (and their solutions) to make progress with _this_ PR. I am not asking to add any TODOs or any controversial descriptions regarding these data members. We can choose to make progress _despite_ their problems and _despite_ our disagreements about their problems/benefits. For a more specific suggestion on how to move forward regarding those counters, please follow that link: https://github.com/squid-cache/squid/pull/1211#discussion_r1050810870

> instead of obtusely requiring design changes.

Please tone down your comments. We need to cooperate (or at least argue) using technical arguments. There is no need to use words that can be easily misinterpreted as personal attacks on others, especially when others are simply answering your question.

Reply:
I want several changes. As far as **this** change request is concerned, I ([still](https://github.com/squid-cache/squid/pull/1211#discussion_r1050802407)) suggest simply removing the second sentence of the method description. To move forward, we do not need to agree on the performance impact of this method (as long as neither of the conflicting opinions is documented as the established fact).